lib/authentication/secure_storage/json_credential_manager.js

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2015-2024 Snowflake Computing Inc. All rights reserved.
+ */
+
+const path = require('path');
+const Logger = require('../../logger');
+const fs = require('fs');
+const os = require('os');
+const Util = require('../../util');
+
+function JsonCredentialManager() {
+
+  this.userinfo = os.userInfo();
+
+  this.getTokenDir = function () {
+    let tokenDir = process.env.SF_TEMPORARY_CREDENTIAL_CACHE_DIR;
+
+    if (!Util.exists(tokenDir)) {
+      tokenDir = os.homedir();
+    } else {
+      Logger.getInstance().info(`The credential cache directory is configured by the user. The token will be saved at ${tokenDir}`);
+    }
+
+    if (!Util.exists(tokenDir)) {
+      Logger.getInstance().error(`Temporary credential cache directory is invalid, and the driver is unable to use the default location(home). 
+      Please assign the enviroment variable value SF_TEMPORARY_CREDENTIAL_CACHE_DIR to enable the default credential manager.`);
+      return null; 
+    }
+    const tokenCacheFile = path.join(tokenDir, 'temporary_credential.json');
+
+    if (fs.existsSync(tokenCacheFile)) {
+      try {
+        fs.accessSync(tokenCacheFile,  fs.constants.R_OK | fs.constants.W_OK);
+      } catch (err) {
+        Logger.getInstance().info(`You do not have permission to the credential file: ${tokenCacheFile}.`);
+        return null;
+      }
+    }
+    Logger.getInstance().info('Cache directory: ', tokenCacheFile);
+    return tokenCacheFile;
+  };
+
+  this.tokenDir = this.getTokenDir();
+
+  this.readJsonCredentialFile = function () {
+    if (this.tokenDir === null) {
+      return null;
+    }
+
+    try {
+      const cred = fs.readFileSync(this.tokenDir, 'utf8');
+      return JSON.parse(cred);
+    } catch (err) {
+      Logger.getInstance().info(`Cannot read token data from the file in ${this.tokenDir}. Please check the permission or the file format of the token.`);
+      return null;
+    }
+  };
+  
+  this.write = function (key, token) {
+    if (!validateTokenCacheOption(key)) {
+      return null;
+    }
+    
+    let jsonCredential = null;
+  
+    if (fs.existsSync(this.tokenDir)){
+      jsonCredential = this.readJsonCredentialFile() || {};
+    } else {
+      jsonCredential = {}; 
+    }
+    jsonCredential[key] = token;
+   
+    try {
+      fs.writeFileSync(this.tokenDir, JSON.stringify(jsonCredential), { mode: 0o600 });
```

Comment:
if file exists after 'writeFileSync' there will be updated data and access 600 even if didn't before?

Reply:
Sorry, It is not clear to me. Could you please elaborate on this?

Reply:
It is connected with fs.accessSync(tokenCacheFile, fs.constants.R_OK | fs.constants.W_OK). It should be checked before writing token to file as well.

Reply:
fs.constants.R_OK | fs.constants.W_OK means the users should have both the read and write permission. Although we already checked before creating the instance, do we need to validate the permission before writing the cache? Or, do you mean we need to check that the users have the permission of fs.constants.S_IRUSR and fs.constants.S_IWUSR?

Reply:
What about: 'fs.statSync('./yourfile.txt').mode'?

Reply:
I added it.